semantic-ast/src/Marshal/JSON.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Marshal.JSON
+( MarshalJSON(..)
+) where
+
+import Data.Aeson as Aeson
+import Data.List.NonEmpty (NonEmpty)
+import GHC.Generics
+import Data.Text (Text)
+import qualified Data.Text as Text
+
+-- Serialize unmarshaled ASTs into JSON representation by auto-deriving Aeson instances generically
+class MarshalJSON t where
+  marshal :: (ToJSON a) => t a -> Value
+  default marshal :: ( Generic1 t, GMarshalJSON (Rep1 t), ToJSON a) => t a -> Value
+  marshal = gmarshal . from1
+
+-- Create MarshalJSON instances for each type constructor
+instance (GMarshalJSON (Rep1 t), Generic1 t) => MarshalJSON t
+
+-- Typeclass to generically marshal ASTs into JSON
+class GMarshalJSON f where
+  gmarshal :: (ToJSON a) => f a -> Value
```

Comment:
Given some type `a` that's an instance of `ToJSON`, we apply a function `f` and return a JSON value represented as a Haskell value